Reaction ensemble maintainance

jngrad opened the pull request:
Description of changes:
- modernize code
- remove unused functions
- unit test helper functions

jngrad committed:
core: Simplify code

jngrad committed:
core: Remove unused functions

jngrad committed:
core: Move utility functions to header file

jngrad committed:
core: Guard against NaN values

jngrad committed:
core: Unit test ReactionEnsemble utility functions

fweik commented on src/core/reaction_ensemble.hpp:
```diff
@@ -389,9 +389,64 @@ class WidomInsertion : public ReactionAlgorithm {
+/**
+ * Checks whether a number is in an array.
+ */
+inline bool is_in_vector(int value, std::vector<int> const &rng) {
```
That is `Utils::contains`

fweik commented on src/core/reaction_ensemble.hpp:
```diff
@@ -389,9 +389,64 @@ class WidomInsertion : public ReactionAlgorithm {
+ * Finds the minimum non negative value in the provided double array and returns
+ * this value.
+ */
+inline double find_minimum_non_negative_value(std::vector<double> const &rng) {
```
This should probably be formulated as an accumulate. This implementation is UB if `rng` is empty, which needs to be checked.

jngrad committed:
core: Use accumulators

fweik reviewed (dismissed):
LGTM

jngrad committed:
Clang-Tidy suggestion

jngrad commented:
coyote 11 with 3 CI jobs running in parallel:
```
45/148 Test #101: wang_landau_reaction_ensemble .............................   Passed  286.36 sec
```
The test is still going to time out even on the most recent machines under a small load. We have to remove it from the coverage.

jngrad committed:
CI: Remove Wang-Landau test from coverage builds

The test times out on CI runners under load when coverage is enabled.
The runtime increases from 220s on a runner without load to 280-300s
when 2 or more CI jobs are executed on the same runner.

fweik commented:
The coverage matters because it has different optimizer flags?

jngrad commented:
This plays a role. Although in `cuda9-maxset`, the test is often 3 times faster. Maybe the GPU tests running at the same time as Wang-Landau take some load off the CPU?

jngrad commented:
here are the relevant optimization flags:
```
cuda10-maxset
-O3 -g -fPIC   -pthread [-Wall and friends...] 
default
-Og -g -fPIC   -pthread [-Wall and friends...] -g --coverage -fprofile-arcs -ftest-coverage
cuda9-maxset
-Og -g -fPIC   -pthread [-Wall and friends...] -g --coverage -fprofile-arcs -ftest-coverage
```
The custom test timeout values in `.gitlab-ci.yml` don't seem to have been updated in a while. For example, `cuda9-maxset` has 900s timeout instead of the default 300s, while its tests take at most 60s.
